clang/test/SemaSYCL/num_simd_work_items.cpp

```diff
@@ -0,0 +1,65 @@
+// RUN: %clang %s -fsyntax-only -fsycl-device-only -DTRIGGER_ERROR -Xclang -verify
+// RUN: %clang %s -fsyntax-only -Xclang -ast-dump -fsycl-device-only | FileCheck %s
+// RUN: %clang_cc1 -fsycl-is-host -fsyntax-only -verify %s
+
+#ifndef __SYCL_DEVICE_ONLY__
+struct FuncObj {
+  [[intelfpga::num_simd_work_items(42)]] // expected-no-diagnostics
+  void operator()() {}
+};
+
+template <typename name, typename Func>
+void kernel(Func kernelFunc) {
+  kernelFunc();
+}
+
+void foo() {
+  kernel<class test_kernel1>(
+      FuncObj());
+}
+
+#else // __SYCL_DEVICE_ONLY__
+
+[[intelfpga::num_simd_work_items(2)]] // expected-warning{{'num_simd_work_items' attribute ignored}}
+void func_ignore() {}
+
+struct FuncObj {
+  [[intelfpga::num_simd_work_items(42)]]
+  void operator()() {}
+};
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  // CHECK-LABEL: FunctionDecl {{.*}} _ZTSZ4mainE12test_kernel1
+  // CHECK:       SYCLIntelNumSimdWorkItemsAttr {{.*}} 42
+  kernel<class test_kernel1>(
+      FuncObj());
+
+  // CHECK-LABEL: FunctionDecl {{.*}} _ZTSZ4mainE12test_kernel2
+  // CHECK:       SYCLIntelNumSimdWorkItemsAttr {{.*}} 8
+  kernel<class test_kernel2>(
+      []() [[intelfpga::num_simd_work_items(8)]] {});
+
+  // CHECK-LABEL: FunctionDecl {{.*}} _ZTSZ4mainE12test_kernel3
+  // CHECK-NOT:   SYCLIntelNumSimdWorkItemsAttr {{.*}} 2
+  kernel<class test_kernel3>(
+      []() {func_ignore();});
+
+#ifdef TRIGGER_ERROR
+  [[intelfpga::num_simd_work_items(0)]] int Var = 0; // expected-error{{'num_simd_work_items' attribute only applies to functions}}
+
+  kernel<class test_kernel4>(
+      []() [[intelfpga::num_simd_work_items(0)]] {}); // expected-error{{'num_simd_work_items' attribute must be greater than 0}}
+
+  kernel<class test_kernel5>(
+      []() [[intelfpga::num_simd_work_items(-42)]] {}); // expected-error{{'num_simd_work_items' attribute requires a non-negative integral compile time constant expression}}
+
+  kernel<class test_kernel6>(
+      []() [[intelfpga::num_simd_work_items(1), intelfpga::num_simd_work_items(2)]] {}); // expected-warning{{attribute 'num_simd_work_items' is already applied with different parameters}}
```

Comment:
I'd also suggest checking for `note: previous attribute is here`.

Also, I don't like the fact that it's not clear which of the two attributes will actually be used.

Reply:
There is no "note" msg expected.
Actually, I started to think making the diagnostic to be error. Want to change it in a different patch tho, since this warning is used somewhere else. What do you think? Also I'd like to hear from @erichkeane

Reply:
1. I agree it's reasonable for this attribute to error out on duplication instead of producing a warning and obfuscating the compiler choice.
2. Instead of doing `warn_*` -> `err_*` for this diagnostic, I'd suggest simply adding an `err_` option and using it for `num_simd_work_items`. This seems more flexible than making all previous diagnostic usages "erroneous" because of the new attribute.

Reply:
Sorry for the confusion, I wan't about changing warn_* to err_*, instead I'd like to add err_* diagnostics and apply it for all SYCL kernel attributes. Still wanr_* will present for non-SYCL (or SYCL non-kernel) attrs.

Reply:
Sounds reasonable, TBD then

Reply:
"Attribute ignored" is typically a warning, no matter the reasoning.  WIth SYCL attributes, I'd like us to simply be consistent between them, and with what the standard says.

Can some one quote the SYCL standard on some attribute about duplicate/conflicting attributes?  Does it say "ill-formed" or something?

Reply:
Spec states (6.7 Attributes):
>  If there are any conflicts
between different kernel attributes, then the behavior is undefined.

So we possible can do whatever we want.

Reply:
Ugg... that is a terrible application of UB.  I'm Ok with whatever we choose, as long as its consistent.